src/main/scala/strawman/collection/immutable/Vector.scala

```diff
@@ -0,0 +1,1140 @@
+    else
+      throw new IndexOutOfBoundsException(index.toString)
+  }
+
+  /*override*/ def updated[B >: A](index: Int, elem: B): Vector[B] = updateAt(index, elem)
+
+  /** Alias for `appendFront`.
+    *
+    * Note that :-ending operators are right associative (see example).
+    * A mnemonic for `+:` vs. `:+` is: the COLon goes on the COLlection side.
+    */
+  @`inline` def +: [B >: A](elem: B): Vector[B] = appendFront(elem)
+
+  /** Alias for `appendBack`
+    *
+    * Note that :-ending operators are right associative (see example).
+    * A mnemonic for `+:` vs. `:+` is: the COLon goes on the COLlection side.
+    */
+  @`inline` def :+ [B >: A](elem: B): Vector[B] = appendBack(elem)
+
+  override def take(n: Int): Vector[A] = {
+    if (n <= 0)
+      Vector.empty
+    else if (startIndex < endIndex - n)
+      dropBack0(startIndex + n)
+    else
+      this
+  }
+
+  override def drop(n: Int): Vector[A] = {
+    if (n <= 0)
+      this
+    else if (startIndex < endIndex - n)
+      dropFront0(startIndex + n)
+    else
+      Vector.empty
+  }
+
+  /*override*/ def takeRight(n: Int): Vector[A] = {
+    if (n <= 0)
+      Vector.empty
+    else if (endIndex - n > startIndex)
+      dropFront0(endIndex - n)
+    else
+      this
+  }
+
+  /*override*/ def dropRight(n: Int): Vector[A] = {
+    if (n <= 0)
+      this
+    else if (endIndex - n > startIndex)
+      dropBack0(endIndex - n)
+    else
+      Vector.empty
+  }
+
+  override def head: A = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.head")
+    apply(0)
+  }
+
+  override def tail: Vector[A] = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.tail")
+    drop(1)
+  }
+
+  override def last: A = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.last")
+    apply(length - 1)
+  }
+
+  /*override*/ def init: Vector[A] = {
+    if (isEmpty) throw new UnsupportedOperationException("empty.init")
+    dropRight(1)
+  }
+
+  override def slice(from: Int, until: Int): Vector[A] =
+    take(until).drop(from)
+
+  override def splitAt(n: Int): (Vector[A], Vector[A]) = (take(n), drop(n))
+
+  // concat (suboptimal but avoids worst performance gotchas)
+//  override def concat[B >: A](that: IterableOnce[B]): Vector[B] = {
+//    import Vector.{Log2ConcatFaster, TinyAppendFaster}
+//    if (that.isEmpty) this
+//    else {
+//      that.size match {
+//        // Often it's better to append small numbers of elements (or prepend if RHS is a vector)
+//        case n if n <= TinyAppendFaster || n < (this.size >>> Log2ConcatFaster) =>
+//          var v: Vector[B] = this
+//          for (x <- that) v = v :+ x
+//          v
+//        case n if this.size < (n >>> Log2ConcatFaster) && that.isInstanceOf[Vector[_]] =>
+//          var v = that.asInstanceOf[Vector[B]]
+//          val ri = this.reverseIterator
+//          while (ri.hasNext) v = ri.next +: v
+//          v
+//        case _ => super.concat(that)
+//      }
+//    }
+//  }
```

Comment:
These optimizations can not be applied in the new design because `concat` is defined in `IterableOps` as a method taking as parameter a `IterableOnce`. This type is extremely limited because we can consume its elements only once… In these optimizations we do it more than once (e.g. `isEmpty` followed by `size`).

One solution would be to change the type signature of `concat` to take an `Iterable` instead of `IterableOnce`. Do you see a reason **not** to do that?